pkg/pow/submitter.go

```diff
@@ -0,0 +1,251 @@
+	"github.com/tellor-io/telliot/pkg/config"
+	"github.com/tellor-io/telliot/pkg/contracts"
+	"github.com/tellor-io/telliot/pkg/db"
+	"github.com/tellor-io/telliot/pkg/rpc"
+	"github.com/tellor-io/telliot/pkg/tracker"
+)
+
+/**
+* The solution handler has one purpose: to either submit the solution on-chain
+* or to reject it if the miner has already submitted a solution for the challenge
+* or the the solution's challenge does not match current challenge
+ */
+
+type ResultData struct {
+	*Result
+	SuccessCallback func(*types.Transaction)
+}
+type Submitter struct {
+	ctx              context.Context
+	close            context.CancelFunc
+	logger           log.Logger
+	cfg              *config.Config
+	proxy            db.DataServerProxy
+	account          *rpc.Account
+	client           contracts.ETHClient
+	tellor           *contracts.ITellor
+	currentChallenge *MiningChallenge
+	currentNonce     string
+	currentValues    [5]*big.Int
+	submitter        tellorCommon.TransactionSubmitter
+	resultCh         chan *ResultData
+	submitCount      prometheus.Counter
+	submitFailCount  prometheus.Counter
+}
+
+func CreateSubmitter(ctx context.Context, cfg *config.Config, logger log.Logger, client contracts.ETHClient, tellor *contracts.ITellor, account *rpc.Account, txSubmitter tellorCommon.TransactionSubmitter, proxy db.DataServerProxy) (*Submitter, chan *ResultData) {
+	ctx, close := context.WithCancel(ctx)
+	submitter := &Submitter{
+		ctx:       ctx,
+		close:     close,
+		proxy:     proxy,
+		submitter: txSubmitter,
+		cfg:       cfg,
+		resultCh:  make(chan *ResultData),
+		account:   account,
+		logger:    log.With(logger, "component", ComponentName),
+		submitCount: promauto.NewCounter(prometheus.CounterOpts{
+			Namespace:   "telliot",
+			Subsystem:   "mining",
+			Name:        "submit_total",
+			Help:        "The total number of submitted solutions",
+			ConstLabels: prometheus.Labels{"account": account.Address.String()},
+		}),
+		submitFailCount: promauto.NewCounter(prometheus.CounterOpts{
+			Namespace:   "telliot",
+			Subsystem:   "mining",
+			Name:        "submit_fails_total",
+			Help:        "The total number of failed submission",
+			ConstLabels: prometheus.Labels{"account": account.Address.String()},
+		}),
+	}
+
+	return submitter, submitter.resultCh
+}
+
+func (s *Submitter) Start() error {
+	for {
+		select {
+		case <-s.ctx.Done():
+			return errors.New("submitter is closing")
+		case result := <-s.resultCh:
+			level.Debug(s.logger).Log("msg", "new result was sent to the submitter", "account", s.account.Address.String())
+			tx, err := s.Submit(s.ctx, result.Result)
+			if err != nil {
+				level.Error(s.logger).Log("msg", "submiting a solution", "err", err, "account", s.account.Address.String())
+				s.submitFailCount.Inc()
+				continue
+			}
+			level.Debug(s.logger).Log("msg", "submited a solution", "txHash", tx.Hash().String(), "account", s.account.Address.String())
+			s.saveGasUsed(s.ctx, tx)
+			s.submitCount.Inc()
```

Comment:
This needs to go in a go routine so that a blocked submit doesn't block the receiving channel. for example imagine the submit call blocks for some reason(infura is hanged) this means that the mining mgr can't send solutions anymore.

Also here are want the logic so that if a new result comes need to send a cancel to all pending  submits.

```
		case result := <-s.resultCh:
			if mgr.LastSubmitCncl !=nil{
				mgr.LastSubmitCncl()
			}
			
			ctx,mgr.LastSubmitCncl = context.WithCancel(s.ctx)
			submit(ctx)
			
....	
function (mgr) submit(ctx,...){
	go func(ctx){
			level.Debug(s.logger).Log("msg", "new result was sent to the submitter", "account", s.account.Address.String())
			tx, err := s.Submit(ctx, result.Result)
			if err != nil {
				level.Error(s.logger).Log("msg", "submiting a solution", "err", err, "account", s.account.Address.String())
				s.submitFailCount.Inc()
				continue
			}
			level.Debug(s.logger).Log("msg", "submited a solution", "txHash", tx.Hash().String(), "account", s.account.Address.String())
			s.saveGasUsed(ctx, tx)
			s.submitCount.Inc()
	}(ctx)
}
```

Reply:
Sorry, why did you mentioned mgr in the snippet? do you mean the submitter?

Reply:
yes submitter